src/products/views/ProductUpdate/handlers/errors.ts

```diff
@@ -41,80 +35,66 @@ export type ProductVariantListError =
       type: "create";
     };
 
-export function getProductVariantListErrors(
-  productChannelsUpdateResult: FetchResult<ProductChannelListingUpdateMutation>,
-  variantMutationResults: FetchResult[],
+export function getCreateVariantMutationError(
+  result: FetchResult<ProductVariantBulkCreateMutation>,
+): ProductVariantListError[] {
+  return result.data.productVariantBulkCreate.errors.map<ProductVariantListError>(
+    error => ({
+      __typename: "DatagridError",
+      type: "create",
+      index: error.index,
+      error: error.code,
+    }),
+  );
+}
+
+export function getVariantUpdateMutationErrors(
+  mutationResult: FetchResult<ProductVariantBulkUpdateMutation>,
+  varaintsIds: string[],
 ): ProductVariantListError[] {
-  return [productChannelsUpdateResult, ...variantMutationResults]
-    .filter(hasMutationErrors)
-    .flatMap(result => {
-      if (result.data.productVariantChannelListingUpdate) {
-        const data = result.data as ProductVariantChannelListingUpdateMutation;
-        return data.productVariantChannelListingUpdate.errors.map<
-          ProductVariantListError
-        >(error => ({
-          __typename: "DatagridError",
-          type: "channel",
-          error: error.code,
-          variantId: (result.extensions
-            .variables as ProductVariantChannelListingUpdateMutationVariables)
-            .id,
-          channelIds: error.channels,
-        }));
-      }
+  const { productVariantBulkUpdate } = mutationResult.data;
+  const generalErrors = productVariantBulkUpdate.errors;
+  const variantsErrors = productVariantBulkUpdate.results.flatMap(
+    res => res.errors,
+  );
+
+  return [...generalErrors, ...variantsErrors].reduce<
```

Comment:
suggestion: we have a big reduce function here that handles multiple cases. What if you can split it into one responsibility each:

```
[...generalErrors, ...variantsErrors]
.reduce(channelErrors)
.reduce(warehouseErrors)
```

Look how readable it becomes. You can create reduces above `getVariantUpdateMutationErrors` or even move it to the separated file. If your reducers have some dependencies (lifted up), create factories for them:

```
const channelErrors = createChannelErrorsReducer(some params)
const warehouseErrors = createWarehouseReducer(some params)


[...generalErrors, ...variantsErrors]
.reduce(channelErrors)
.reduce(warehouseErrors)
```

or simpler:

```


[...generalErrors, ...variantsErrors]
.reduce(channelErrors(some params))
.reduce(warehouseErrors(some params))
```